site/js/ModalsActor.js

```diff
@@ -0,0 +1,184 @@
+    modalEl.classList.add('modal')
+
+    const titleEl = document.createElement('h2')
+    titleEl.classList.add('modal-title')
+    titleEl.appendChild(document.createTextNode(modalData.title))
+
+    modalEl.appendChild(titleEl)
+
+    if (modalData.desc) {
+      const descEl = document.createElement('p')
+      descEl.classList.add('modal-desc')
+      descEl.appendChild(document.createTextNode(modalData.desc))
+
+      modalEl.appendChild(descEl)
+    }
+
+    let errorEl, inputEl
+    if (modalData.input) {
+      errorEl = document.createElement('span')
+      errorEl.classList.add('modal-error')
+
+      modalEl.appendChild(errorEl)
+
+      inputEl = document.createElement('input')
+      inputEl.classList.add('modal-input')
+      inputEl.classList.add('input')
+      inputEl.placeholder = modalData.input.placeholder || ''
+      inputEl.type = modalData.input.type || 'text'
+
+      inputEl.addEventListener('keydown', evt => {
+        if (evt.keyCode === 13) {
+          // Return/enter key, submit form
+          const btns = modalData.buttons.filter(btn => btn.doValidate === true)
+          
+          if (btns.length === 1) {
+            btns[0].el.click()
+          } else if (btns.length > 1) {
+            // There are multiple options available, so don't
+            // assume one!
+          }
+        }
+      })
+
+      modalEl.appendChild(inputEl)
+    }
+
+    const actionsEl = document.createElement('div')
+    actionsEl.classList.add('modal-actions')
+    for (const btnData of modalData.buttons) {
+      const btnEl = document.createElement('button')
+      btnEl.classList.add('btn')
+
+      if (btnData.primary) {
+        btnEl.classList.add('btn-primary')
+      }
+
+      btnData.el = btnEl
+      btnEl.appendChild(document.createTextNode(btnData.text))
+
+      btnEl.addEventListener('click', async evt => {
+        evt.preventDefault()
+
+        // Validate form
+        if (inputEl && modalData.input.validateFn && btnData.doValidate) {
+          try {
+            await modalData.input.validateFn(inputEl.value || inputEl.placeholder)
+          } catch (error) {
+            // Display error
+
+            errorEl.innerText = error
+            this.emit('modal validation error', error)
+
+            return
+          }
+        }
+
+        // Close the modal
+        modalEl.remove()
+        this.emit('close modal', modalData, modalEl)
+        document.getElementById('app').classList.remove('modal-visible')
+
+        // Trigger callback
+        if (inputEl) {
+          if (inputEl.value.length === 0) {
+            // Use placeholder as default
+            btnData.onclick(inputEl.placeholder)
+          } else {
+            btnData.onclick(inputEl.value)
+          }
+        } else {
+          btnData.onclick(btnData.text)
+        }
+
+        // Continue displaying modals from the backlog
+        this.activeModal = null
+        this.displayModalsFromBacklog()
```

Comment:
Good thinking!

Reply:
Yup! This allows you to basically assign a bunch of modals and Promise.all them:

```
const [ username, email ] = await Promise.all([
  this.prompt('username pls'),
  this.prompt('email pls'),
])
```

We don't actually do this yet though 📦